gninasrc/lib/flexinfo.cpp

```diff
@@ -102,6 +103,76 @@ void FlexInfo::sanitizeResidues(OpenBabel::OBMol& receptor) {
   }
 }
 
+void FlexInfo::keepNearestResidues(
+  const OpenBabel::OBMol& rigid, 
+  std::unordered_map<std::size_t, double>& residues_distances){
+
+  using namespace OpenBabel;
+
+  // Loop over residue list and compute minimum distances
+  for(auto& resdist : residues_distances){
+
+    // Current residue (among flexible ones)
+    std::size_t residx = resdist.first; // Key
+    OBResidue* res = rigid.GetResidue(residx);
+
+    // Minimum distance between ligand and current residue atoms
+    double d2;
+
+    // Loop over ligand atoms
+    FOR_ATOMS_OF_MOL(alig, distligand)
+    {
+      vector3 al = alig->GetVector();
+
+      // Loop over residue atoms
+      for(const auto ares: res->GetAtoms()){
+
+        if(!isSideChain(ares->GetResidue()->GetAtomID(&(*ares)))) continue; // skip backbone
```

Comment:
I'm not very happy with `ares->GetResidue()->GetAtomID(&(*ares)))`. Is there a better way to get the PDB atom names (stored in `OBResidue`) from an `OBAtom`? I might have overlooked something in the documentation, but I couldn't find a better way of achieving this.

Reply:
This line of code is nastier than I previously thought by running only the docking; many `gnina` test fail with this line but work when it's commented out (together with the other similar line)... Any suggestion on how to do this properly?

Reply:
The residue is where the atom names are.  There's no need for &(*ares) - can just do ares.

What is the failure, is ares sometimes null?

Reply:
It was a segmentation fault that I could not elucidate with `valgrind` (I'll check if it's `null`). Here I changed it to `res->GetAtomID(ares)` since I'm looping over residues first and this solves the issue. 

However, where I'm looping over atoms and I still need to use `a->GetResidue()->GetAtomID(&(*a))` (which still causes problems), and using `a->GetResidue()->GetAtomID(a)` as you suggest gives the following problem:

```
/biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.cpp: In member function 'void FlexInfo::extractFlex(OpenBabel::OBMol&, OpenBabel::OBMol&, std::__cxx11::string&)':
/biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.cpp:192:49: error: no matching function for call to 'OpenBabel::OBResidue::GetAtomID(OpenBabel::OBMolAtomIter&)'
     if(!isSideChain(a->GetResidue()->GetAtomID(a))) continue; // skip backbone
                                                 ^
In file included from /usr/local/include/openbabel3/openbabel/atom.h:34:0,
                 from /biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/box.h:3,
                 from /biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.h:6,
                 from /biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.cpp:1:
/usr/local/include/openbabel3/openbabel/residue.h:127:20: note: candidate: std::__cxx11::string OpenBabel::OBResidue::GetAtomID(OpenBabel::OBAtom*) const
     std::string    GetAtomID(OBAtom *atom)        const;
                    ^~~~~~~~~
/usr/local/include/openbabel3/openbabel/residue.h:127:20: note:   no known conversion for argument 1 from 'OpenBabel::OBMolAtomIter' to 'OpenBabel::OBAtom*'
```

Reply:
Oh right, if it is an iterator and not a pointer you do need that ugliness..

Reply:
`a->GetResidue()` is indeed `NULL` some times; I'll investigate further, thanks for the pointer ([xkcd.com/138/](https://www.xkcd.com/138/))!